rust/src/model/mod.rs

```diff
@@ -0,0 +1,191 @@
+pub mod date;
+pub mod duration;
+
+use rand::prelude::*;
+use rocket::serde::Serialize;
+use scylla::cql_to_rust::{FromCqlVal, FromCqlValError};
+use scylla::frame::response::result::CqlValue;
+use scylla::frame::value::{Value, ValueTooBig};
+use scylla::FromRow;
+use scylla::{BufMut, ValueList};
+use struct_field_names::StructFieldNames;
+use struct_field_names_as_array::FieldNamesAsArray;
+use uuid::Uuid;
+
+use crate::date::Date;
+use crate::duration::Duration;
+
+#[derive(Clone, ValueList, FromRow, Serialize, StructFieldNames, FieldNamesAsArray)]
+pub struct Owner {
+    pub owner_id: Uuid,
+    pub address: String,
+    pub name: String,
+}
+
+impl Owner {
+    pub fn random() -> Self {
+        Self {
+            owner_id: Uuid::new_v4(),
+            address: "home".into(),
+            name: random_string(8),
+        }
+    }
+}
+
+#[derive(
+    Debug, Default, Clone, ValueList, FromRow, Serialize, StructFieldNames, FieldNamesAsArray,
+)]
+pub struct Pet {
```

Comment:
When the database is seeded, empty string values exist for pets, the driver seems to be unable to deserialize them back into a struct that has either an `Option<String>` or just a `String`. Not sure what the best way to handle that is since it's a pretty fundamental issue unless I messed it up somewhere.

Reply:
I did a few local tests and a struct with `Option<String>` worked just fine in accepting both nulls and empty strings, as expected.

However, I think I realize what the problem is. The `FromRow` macro does not match the names of attributes, but instead it works **only** on the order of fields.

So, given that the table is defined like that:
```cql
CREATE TABLE IF NOT EXISTS carepet.pet
(
    owner_id UUID,
    pet_id   UUID,
    chip_id  TEXT,
    species  TEXT,
    breed    TEXT,
    color    TEXT,
    gender   TEXT,
    age      INT,
    weight   FLOAT,
    address  TEXT,
    name     TEXT,
    PRIMARY KEY (owner_id, pet_id)
);
```
, that's the same order you should use in the struct in order for it to be deserialized properly:

```rust
    pub owner_id: Uuid,
    pub pet_id: Uuid,
    pub chip_id: Option<String>,
    pub species: Option<String>,
    pub breed: Option<String>,
    pub color: Option<String>,
    pub gender: Option<String>,
    pub age: i32,
    pub weight: f32,
    pub address: String,
    pub name: String,
```

@gngeorgiev please give it a try and let me know if it works.

Ref: https://github.com/scylladb/scylla-rust-driver/issues/370

Reply:
@psarna Thank you for looking into it. Initially, I had the fields ordered in the order you suggested, I moved the species field to be on top to make it more obvious that it's the offending field as I was getting an error for the field with index 3, which is the error I am getting now, after modifying the struct back to its original order. I have commited the new struct order but as mentioned the error persists. Please try following the instructions bellow, I hope that will help you reproduce it:

1. `docker run   -p 9042:9042/tcp   --name scylla   --hostname scylla   -d scylladb/scylla:latest    --smp 1`
2. `rustc 1.57.0 (f1edd0429 2021-11-29)`
3. 
```cql

CREATE KEYSPACE carepet WITH replication = {'class': 'NetworkTopologyStrategy', 'datacenter1': '1'}  AND durable_writes = true;

CREATE TABLE carepet.owner (
    owner_id uuid PRIMARY KEY,
    address text,
    name text
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.pet (
    owner_id uuid,
    pet_id uuid,
    address text,
    age int,
    breed text,
    chip_id text,
    color text,
    gender text,
    name text,
    species text,
    weight float,
    PRIMARY KEY (owner_id, pet_id)
) WITH CLUSTERING ORDER BY (pet_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.sensor (
    pet_id uuid,
    sensor_id uuid,
    type text,
    PRIMARY KEY (pet_id, sensor_id)
) WITH CLUSTERING ORDER BY (sensor_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.sensor_avg (
    sensor_id uuid,
    date date,
    hour int,
    value float,
    PRIMARY KEY (sensor_id, date, hour)
) WITH CLUSTERING ORDER BY (date ASC, hour ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE carepet.measurement (
    sensor_id uuid,
    ts timestamp,
    value float,
    PRIMARY KEY (sensor_id, ts)
) WITH CLUSTERING ORDER BY (ts ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';


INSERT INTO carepet.owner (owner_id, address, name) VALUES (322175a4-9891-4f30-84e0-767361880cd0, 'home', 'vcdpirmb');
INSERT INTO carepet.pet (owner_id, pet_id, address, age, breed, chip_id, color, gender, name, species, weight) VALUES (322175a4-9891-4f30-84e0-767361880cd0, 6b10cefb-2f09-4df3-8659-86e5588f9dad, 'home', 6, null, null, null, null, 'ouethhzv', null, 7.75718);
```
4. `cqlsh -f db.cql`
5. `cargo run`
6. `curl --request GET \
  --url http://127.0.0.1:8000/api/owner/322175a4-9891-4f30-84e0-767361880cd0/pets`

You should get: `{"error":"Bad CQL type in the column with index 3"}`

Reply:
Ugh, it's still about the order, it's just much more confusing than I initially suspected... let me explain more thoroughly. The way Cassandra (and Scylla) treats the order of columns is as follows:
1. partition key and clustering key are always created in order they appeared in the definition
2. static columns go next
3. regular columns come last, in the **lexicographical** order, regardless of the order in which they appeared in the CREATE TABLE statement.

You can see that by running a `DESCRIBE` command in cqlsh. Here's the output for the `pet` table:
```cql
cassandra@cqlsh> describe carepet.pet;

CREATE TABLE carepet.pet (
    owner_id uuid,
    pet_id uuid,
    address text,
    age int,
    breed text,
    chip_id text,
    color text,
    gender text,
    name text,
    species text,
    weight float,
    PRIMARY KEY (owner_id, pet_id)
) WITH CLUSTERING ORDER BY (pet_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

So the actual order for which your code will work (I verified this time, because your reproducing steps worked, thanks!) is as follows - I present it below in form of git diff:
```diff
diff --git a/rust/src/model/mod.rs b/rust/src/model/mod.rs
index 193f91d..d625edb 100644
--- a/rust/src/model/mod.rs
+++ b/rust/src/model/mod.rs
@@ -38,15 +38,15 @@ impl Owner {
 pub struct Pet {
     pub owner_id: Uuid,
     pub pet_id: Uuid,
-    pub chip_id: Option<String>,
-    pub species: Option<String>,
+    pub address: String,
+    pub age: i32,
     pub breed: Option<String>,
+    pub chip_id: Option<String>,
     pub color: Option<String>,
     pub gender: Option<String>,
-    pub age: i32,
-    pub weight: f32,
-    pub address: String,
     pub name: String,
+    pub species: Option<String>,
+    pub weight: f32,
 }
 
 impl Pet {
```

I agree that it's beyond confusing and it makes `FromRow` structs quite hard to grasp and use correctly - perhaps we should indeed consider manufacturing another macro which matches column names to attribute names, to avoid further confusion. /cc @piodul

Reply:
Alternatively, if you explicitly specify all the columns in the query (i.e. `SELECT owner_id, pet_id, chip_id, species, (...)` instead of just `SELECT *`), the values will be fetched in the order that you specified

Reply:
Thanks @psarna, this did fix it 🙂

Reply:
To clarify, I just changed the SELECTs to be explicit, since depending on the ordering of the fields is far from optimal IMO and is not future proof.

Reply:
I agree, an explicit select sounds better